Adding base runtime image for training

Fiona-Waters opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
[RHOAIENG-13188](https://issues.redhat.com/browse/RHOAIENG-13188)
[RHOAIENG-12082](https://issues.redhat.com/browse/RHOAIENG-12082) 

Adding base runtime image with UBI9, CUDA and dependencies needed for fms-hf-tuning, kubeflow training operator, instruct lab and  GPTDolomite.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Testing by running as base image in Dockerfile [here](https://github.com/foundation-model-stack/fms-hf-tuning/tree/main/build) and [here](https://github.com/kubeflow/training-operator/tree/master/sdk/python/kubeflow/trainer). (This is not complete yet).
** at the moment this works in training-operator to the point of running the entry point script where it requires parameters - continuing to look into this.
** Seems not as straightforward for fms-hf-tuning, continuing to look into this.

## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

astefanutti commented on images/runtime/training/cuda/requirements.txt:
```diff
@@ -0,0 +1,34 @@
+peft>=0.8.0,<0.13
```
Would it be possible to move to a Pipefile.lock so builds are reproducible?

Fiona-Waters replied:
I've updated to use Pipfile.lock and added the following as per the ray cuda image
```
# Install micropipenv to deploy packages from Pipfile.lock
RUN pip install --no-cache-dir -U "micropipenv[toml]"

# Install Python dependencies from Pipfile.lock file
COPY Pipfile.lock ./

RUN micropipenv install && rm -f ./Pipfile.lock
```
Unfortunately I am getting an error about the requests dependency which I haven't been able to resolve yet. I tried removing it earlier so that it does not have to uninstall before reinstalling. Will continue to look into it but any advice would be appreciated. This is the error:
```
692.8 Collecting requests==2.32.3 (from -r /tmp/requirements_micropipenv-nhdnz6nv.txt (line 2))
692.8   Using cached requests-2.32.3-py3-none-any.whl (64 kB)
692.8 Installing collected packages: requests
692.8   Attempting uninstall: requests
692.8     Found existing installation: requests 2.25.1
692.8 error: uninstall-no-record-file
692.8 
692.8 × Cannot uninstall requests 2.25.1
692.8 ╰─> The package's contents are unknown: no RECORD file was found for requests.
692.8 
692.8 hint: The package was installed by rpm. You should check if it can uninstall the package.
692.9 Failed to install requirements, dependency 'requests' could not be installed
```

sutaakar reviewed (commented):
/lgtm

astefanutti commented:
/lgtm

astefanutti commented:
We can go without Flash Attention for that first iteration and add it later.

@Fiona-Waters please give us the green light when it's good to go on your side.

Fiona-Waters commented:
@astefanutti @sutaakar I think this is ready to be merged now. Please take one more look over it - since the last review I have updated the labels in the dockerfile and added the async-timeout dependency. Thank you both for all of your help on this.

astefanutti commented on images/runtime/training/cuda/README.md:
```diff
@@ -0,0 +1,8 @@
+It includes the following layers:
+* UBI 9
+* Python 3.11
+* CUDA 12.1
```
Let's add PyTorch there?

Fiona-Waters replied:
Updated, thank you.

Fiona-Waters replied:
Updated, thank you.

astefanutti commented on images/runtime/training/cuda/Pipfile:
```diff
@@ -0,0 +1,34 @@
+transformers = ">=4.41.2"
+numpy = "<2.0.0,>=1.23.5"
+accelerate = ">=0.34.2"
+torch = ">=2.3.0a0"
```
I'd be inclined to pin the PyTorch version as it'll be used in the image tag, WDYT?

Fiona-Waters replied:
Updated, thank you.

sutaakar reviewed (commented):
/lgtm

astefanutti reviewed (approved):
Great work!
